core/templates/dev/head/domain/topic_viewer/read-only-topic-object.factory.ts

```diff
@@ -0,0 +1,155 @@
+import { Subtopic, SubtopicObjectFactory } from
+  'domain/topic/SubtopicObjectFactory';
+
+export interface ISkillDescriptions {
+  [skillId: string]: string | null;
+}
+
+export interface IDegreesOfMastery {
+  [skillId: string]: number | null;
+}
+
+export class ReadOnlyTopic {
+  _topicName: string;
+  _topicId: string;
+  _canonicalStories: Array<StorySummary>;
+  _additionalStories: Array<StorySummary>;
+  _uncategorizedSkills: Array<SkillSummary>;
+  _subtopics: Array<Subtopic>;
+  _degreesOfMastery: IDegreesOfMastery;
+  _skillDescriptions: ISkillDescriptions;
+
+  constructor(topicName: string, topicId: string,
+      canonicalStories: Array<StorySummary>,
+      additionalStories: Array<StorySummary>,
+      uncategorizedSkills: Array<SkillSummary>, subtopics: Array<Subtopic>,
+      degreesOfMastery: IDegreesOfMastery,
+      skillDescriptions: ISkillDescriptions) {
+    this._topicName = topicName;
+    this._topicId = topicId;
+    this._canonicalStories = canonicalStories;
+    this._additionalStories = additionalStories;
+    this._uncategorizedSkills = uncategorizedSkills;
+    this._subtopics = subtopics;
+    this._degreesOfMastery = degreesOfMastery;
+    this._skillDescriptions = skillDescriptions;
+  }
+
+  getTopicName(): string {
+    return this._topicName;
+  }
+
+  getTopicId(): string {
+    return this._topicId;
+  }
+
+  getCanonicalStories(): Array<StorySummary> {
+    return this._canonicalStories.slice();
+  }
+
+  getAdditionalStories(): Array<StorySummary> {
+    return this._additionalStories.slice();
+  }
+
+  getUncategorizedSkills(): Array<SkillSummary> {
+    return this._uncategorizedSkills.slice();
+  }
+
+  getSubtopics(): Array<Subtopic> {
+    return this._subtopics.slice();
+  }
+
+  getDegreesOfMastery(): IDegreesOfMastery {
+    return this._degreesOfMastery;
+  }
+
+  getSkillDescriptions(): ISkillDescriptions {
+    return this._skillDescriptions;
+  }
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+
+export class ReadOnlyTopicObjectFactory {
+  _subtopicObjectFactory: SubtopicObjectFactory;
+  _skillSummaryObjectFactory: SkillSummaryObjectFactory;
+
+  constructor(
+    private subtopicObjectFactory: SubtopicObjectFactory,
+    private skillSummaryObjectFactory: SkillSummaryObjectFactory) {
+    this._subtopicObjectFactory = subtopicObjectFactory;
+    this._skillSummaryObjectFactory = skillSummaryObjectFactory;
+  }
+
+  getStorySummaryArray(storyDicts: any): Array<StorySummary> {
+    var storyReferenceObjectFactory = new StoryReferenceObjectFactory();
+    let storySummaryArray: Array<StorySummary> =
+    storyDicts.map((storyDict: any) => {
+      return new StorySummary( storyDict.id, storyDict.title,
+        storyDict.node_count, storyDict.description,
+        storyReferenceObjectFactory.createFromStoryId(storyDict.id).
+          isStoryPublished());
+    });
+    return storySummaryArray;
+  }
+
+  createFromBackendDict(topicDataDict: any): ReadOnlyTopic {
```

Comment:
Ditto

Reply:
This is actual dict used to created object, can't find a better type. I think I should add comment to remove the `any` ?